### PR TITLE
vmm: add argument --dirty-log

### PR DESCRIPTION
**Pull request by wfly1998**
Current memory_manager will always enable `KVM_MEM_LOG_DIRTY_PAGES` flag, this flag will prevent KVM from using huge pages.

But without this flag, VMM will never migration.

For above reasons, this flag should be an optional parameter for users to choose.

ref:
- [linux/arch/x86/kvm/mmu/mmu.c,L875](https://github.com/torvalds/linux/blob/4b93c544e90e2b28326182d31ee008eb80e02074/arch/x86/kvm/mmu/mmu.c#L875)
- [linux/arch/x86/kvm/mmu/mmu.c,L2886](https://github.com/torvalds/linux/blob/4b93c544e90e2b28326182d31ee008eb80e02074/arch/x86/kvm/mmu/mmu.c#L2886)

```c
static struct kvm_memory_slot *
gfn_to_memslot_dirty_bitmap(struct kvm_vcpu *vcpu, gfn_t gfn,
			    bool no_dirty_log)
{
	struct kvm_memory_slot *slot;

	slot = kvm_vcpu_gfn_to_memslot(vcpu, gfn);
	if (!slot || slot->flags & KVM_MEMSLOT_INVALID)
		return NULL;
	if (no_dirty_log && kvm_slot_dirty_track_enabled(slot))                     // NOTICE HERE
		return NULL;

	return slot;
}

int kvm_mmu_hugepage_adjust(struct kvm_vcpu *vcpu, gfn_t gfn,
			    int max_level, kvm_pfn_t *pfnp,
			    bool huge_page_disallowed, int *req_level)
{
	struct kvm_memory_slot *slot;
	kvm_pfn_t pfn = *pfnp;
	kvm_pfn_t mask;
	int level;

	*req_level = PG_LEVEL_4K;

	if (unlikely(max_level == PG_LEVEL_4K))
		return PG_LEVEL_4K;

	if (is_error_noslot_pfn(pfn) || kvm_is_reserved_pfn(pfn))
		return PG_LEVEL_4K;

	slot = gfn_to_memslot_dirty_bitmap(vcpu, gfn, true);                        // NOTICE HERE
	if (!slot)
		return PG_LEVEL_4K;

	level = kvm_mmu_max_mapping_level(vcpu->kvm, slot, gfn, pfn, max_level);
	if (level == PG_LEVEL_4K)
		return level;

	*req_level = level = min(level, max_level);

	/*
	 * Enforce the iTLB multihit workaround after capturing the requested
	 * level, which will be used to do precise, accurate accounting.
	 */
	if (huge_page_disallowed)
		return PG_LEVEL_4K;

	/*
	 * mmu_notifier_retry() was successful and mmu_lock is held, so
	 * the pmd can't be split from under us.
	 */
	mask = KVM_PAGES_PER_HPAGE(level) - 1;
	VM_BUG_ON((gfn & mask) != (pfn & mask));
	*pfnp = pfn & ~mask;

	return level;
}
```

### CODE REVIEW

**Comment by liuw** [comment]
I don't think adding this option is the right approach.

There is already code to dynamically turning on and off log dirty. Does that not work for you?

File: vmm/src/device_manager.rs [hunk 1]
```diff
@@ -1807,8 +1807,8 @@ impl DeviceManager {
 
         Ok(Arc::new(Console {
             serial,
-            input,
             console_resizer,
+            input,
```
Comment (liuw) [inline vmm/src/device_manager.rs:1811]: Unrelated change here. Please fix this sort of issues for future submissions.

File: vmm/src/memory_manager.rs [hunk 2]
```diff
@@ -749,7 +750,7 @@ impl MemoryManager {
             .ok_or(Error::AllocateMmioAddress)?;
 
         #[cfg(not(feature = "tdx"))]
-        let log_dirty = true;
```
Comment (liuw) [inline vmm/src/memory_manager.rs:752]: Is this what is stopping huge pages from being used? What is you just flip the default to false?

**Comment by rbradford** [comment]
I think the issue are such:

1. I'm surprised if huge pages does not work with migration (@likebreath can you look into that - including seeing what QEMU can do.)
2. I don't think this should be an option. If the user chooses use huge pages (either by `--memory` or `--memory-zone`) then we should simply disable live migration with an error.

File: vmm/src/memory_manager.rs (hunk 2 above)
Comment (wfly1998) [reply vmm/src/memory_manager.rs:752]: The default value is still true.

**Comment by likebreath** [comment]
> I think the issue are such:
> 
> 1. I'm surprised if huge pages does not work with migration (@likebreath can you look into that - including seeing what QEMU can do.)
> 2. I don't think this should be an option. If the user chooses use huge pages (either by `--memory` or `--memory-zone`) then we should simply disable live migration with an error.

Of course. Will do and come back with my findings.

**Comment by wfly1998** [comment]
> I think the issue are such:
> 
> 1. I'm surprised if huge pages does not work with migration (@likebreath can you look into that - including seeing what QEMU can do.)
> 2. I don't think this should be an option. If the user chooses use huge pages (either by `--memory` or `--memory-zone`) then we should simply disable live migration with an error.

Thanks for your review. 

I wanted to disable live migration when huge pages enabled at first, but VMM can still use mounted hugetlbfs as memory backend in `--memory-zone` without declaring `hugepages=on`.

I finally decided to add this option although it’s not elegant enough.

**Comment by likebreath** [comment]
@wfly1998 Thank you for reporting the issue. Can you please elaborate more about the failing behaviors you observed ? Did you see any sort of errors when migrating Cloud Hypervisor with hugepages enabled? Or did you believe there is a silent bug/misbehavior (after reading the code)?

The context is that I did experiment live-migration with the hugepages options  (with `memory-zone` and `numa` options) previously, and the migration was completed "successfully".  I want to make sure the behavior we saw is consistent and we are looking at the same problem. Thank you.

FYI, I was using the following CLI: 
```
/home/chenb/project/cloud-hypervisor/cloud-hypervisor/target/release/cloud-hypervisor --kernel vmlinux --pmem file=focal-server-cloudimg-amd64-custom-20210407-0.raw,discard_writes=on --disk path=tmp.img --cpus boot=6,max=12 --memory size=0 --memory-zone id=mem0,size=256M,hugepages=on id=mem1,size=512M,hugepages=on id=mem2,size=1G,hugepages=on --numa guest_numa_id=0,cpus=0-2:9,distances=1@15:2@20,memory_zones=mem0 guest_numa_id=1,cpus=3-4:6-8,distances=0@20:2@25,memory_zones=mem1 guest_numa_id=2,cpus=5:10-11,distances=0@25:1@30,memory_zones=mem2 --cmdline root=/dev/pmem0p1 console=ttyS0 --net tap=,mac=12:34:56:78:90:01,ip=192.168.1.1,mask=255.255.255.0 --serial tty --console off --api-socket=/tmp/api1 --seccomp false --log-file src.log -v
```

**Comment by wfly1998** [comment]
> @wfly1998 Thank you for reporting the issue. Can you please elaborate more about the failing behaviors you observed ? Did you see any sort of errors when migrating Cloud Hypervisor with hugepages enabled? Or did you believe there is a silent bug/misbehavior (after reading the code)?
> 
> The context is that I did experiment live-migration with the hugepages options (with `memory-zone` and `numa` options) previously, and the migration was completed "successfully". I want to make sure the behavior we saw is consistent and we are looking at the same problem. Thank you.
> 
> FYI, I was using the following CLI:
> 
> ```
> /home/chenb/project/cloud-hypervisor/cloud-hypervisor/target/release/cloud-hypervisor --kernel vmlinux --pmem file=focal-server-cloudimg-amd64-custom-20210407-0.raw,discard_writes=on --disk path=tmp.img --cpus boot=6,max=12 --memory size=0 --memory-zone id=mem0,size=256M,hugepages=on id=mem1,size=512M,hugepages=on id=mem2,size=1G,hugepages=on --numa guest_numa_id=0,cpus=0-2:9,distances=1@15:2@20,memory_zones=mem0 guest_numa_id=1,cpus=3-4:6-8,distances=0@20:2@25,memory_zones=mem1 guest_numa_id=2,cpus=5:10-11,distances=0@25:1@30,memory_zones=mem2 --cmdline root=/dev/pmem0p1 console=ttyS0 --net tap=,mac=12:34:56:78:90:01,ip=192.168.1.1,mask=255.255.255.0 --serial tty --console off --api-socket=/tmp/api1 --seccomp false --log-file src.log -v
> ```

Huge pages and live migration are not in conflict in current cloud hypervisor, because vmm is really using huge pages, but kvm find that dirty log is enabled, then kvm build extended page table with 4K page. 

In other words, vmm cost huge pages but kvm treated them as simple 4K pages silently. 

It’s really difficult to find, but it’s easy to bench. 

You can simply use `perf` to check pmu-events such as `ept.walk_cycles` or `ept.walk_pending` with huge pages enabled or disabled, then check it again with flipping  `log_dirty` to false and huge pages enabled. 

It’s obvious.

**Comment by wfly1998** [comment]
What’s more, please guarantee that extended page tables (EPT in Intel VT-x) is enabled.

**Comment by likebreath** [comment]
@wfly1998 Appreciate your detailed explanation. Good to confirm it is a matter of silent misbehavior as you observed. Let me take a closer look and will get back to it later.

**Comment by likebreath** [comment]
After poking around QEMU code a while (mostly code from [`migration`](https://github.com/qemu/qemu/tree/master/migration) and [`accel/kvm`](https://github.com/qemu/qemu/tree/master/accel/kvm) folder), I didn't find special handling for huge-pages with live-migration. So I started to look into the KVM code from the Linux kernel related to `KVM_MEM_LOG_DIRTY_PAGES`. As you pointed earlier, I did confirm that the granule of KVM dirty log is 4K. 

If I read the KVM code correctly [1], I think the guest memory with huge-pages should be correctly handled by the current live-migration support in Cloud Hypervisor.

1. The `KVM_MEM_LOG_DIRTY_PAGES` flag is only enabled during live-migration [2] and is turned off after live-migration [3]. Note all `MemoryRegion` are constructed without "dirty_logging" [4].
2. The KVM will drop the splitted small-page spte when "diry logging" is stopped (so that large-page sptes can be re-created) [5]. Quote: 
```
		 * Scan sptes if dirty logging has been stopped, dropping those
		 * which can be collapsed into a single large-page spte.  Later
		 * page faults will create the large-page sptes.
```

[1] https://github.com/torvalds/linux/blob/ac08b1c68d1b1ed3cebb218fc3ea2c07484eb07d/arch/x86/kvm/x86.c#L11534
[2] https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/hypervisor/src/kvm/mod.rs#L435
[3] https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/hypervisor/src/kvm/mod.rs#L459
[4] https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/hypervisor/src/kvm/mod.rs#L302
[5] https://github.com/torvalds/linux/blob/ac08b1c68d1b1ed3cebb218fc3ea2c07484eb07d/arch/x86/kvm/x86.c#L11584-L11586

I may miss some context about the KVM code from the Linux kernel. Please let me know if you have any questions or comments. Thank you.

**Comment by likebreath** [comment]
On the other hand, I read from QEMU's migration document [1] that it requires `userfault` support from the destination machine as described in the `Postcopy with hugepages` section. I am not sure whether this is something we will also need for live-migration with Cloud Hypervisor.

[1] https://github.com/qemu/qemu/blob/master/docs/devel/migration.rst

**Comment by liuw** [comment]
> On the other hand, I read from QEMU's migration document [1] that it requires `userfault` support from the destination machine as described in the `Postcopy with hugepages` section. I am not sure whether this is something we will also need for live-migration with Cloud Hypervisor.
> 
> [1] https://github.com/qemu/qemu/blob/master/docs/devel/migration.rst

Cloud Hypervisor is doing pre-copy not post-copy, right? I think this question is moot?

**Comment by wfly1998** [comment]
@likebreath Thanks for your explanation, it's my fault.
All of my tests was based on my own fork without recent commits, which caused this problem.
I'm so sorry for wasting your time, maybe I should test on latest version or open an issue rather than open pr directly.
Thank you again for your patient explanation.

**Comment by likebreath** [comment]
> Cloud Hypervisor is doing pre-copy not post-copy, right? I think this question is moot?

@liuw Thanks for your comments. That's where I was missing some context. Would you please explain (or share some pointers to) what is pre-copy and post-copy in the context of live-migration?

**Comment by liuw** [comment]
> > Cloud Hypervisor is doing pre-copy not post-copy, right? I think this question is moot?
> 
> @liuw Thanks for your comments. That's where I was missing some context. Would you please explain (or share some pointers to) what is pre-copy and post-copy in the context of live-migration?

Pre-copy is you copy as much as you can until the dirty set is below a certain threshold, then you pause the VM and transfer the rest of the state.

Post-copy is you copy as little as you can, and transfer state as the source VM runs.

They are just different strategies for live migration. Cloud Hypervisor's current implementation is pre-copy if I'm not mistaken.

**Comment by likebreath** [comment]
> Pre-copy is you copy as much as you can until the dirty set is below a certain threshold, then you pause the VM and transfer the rest of the state.
> 
> Post-copy is you copy as little as you can, and transfer state as the source VM runs.
> 
> They are just different strategies for live migration. Cloud Hypervisor's current implementation is pre-copy if I'm not mistaken.

Yes, that is right. Appreciate the explanation.

**Comment by likebreath** [comment]
@wfly1998 No problem. It is also a good learning experience to me. 

Thanks for your interest in Cloud Hypervisor. Yes, issues (and also discussions) are great way to start quick conversations if you have any further questions or suggestions. Look forward to talking with you more.

**Pull request closed without merging**